cluster_pack/uploader.py

```diff
@@ -174,32 +185,49 @@ def _get_hash(spec_file: str) -> str:
 
 def _upload_zip(
     zip_file: str, package_path: str,
-    resolved_fs: Any = None, force_upload: bool = False
+    resolved_fs: filesystem.EnhancedFileSystem, force_upload: bool = False
 ) -> None:
     packer = packaging.detect_packer_from_file(zip_file)
     if packer == packaging.PEX_PACKER and resolved_fs.exists(package_path):
-        with tempfile.TemporaryDirectory() as tempdir:
-            local_copy_path = os.path.join(tempdir, os.path.basename(package_path))
-            resolved_fs.get(package_path, local_copy_path)
-            info_from_storage = PexInfo.from_pex(local_copy_path)
-            into_to_upload = PexInfo.from_pex(zip_file)
-            if not force_upload and info_from_storage.code_hash == into_to_upload.code_hash:
-                _logger.info(f"skip upload of current {zip_file}"
-                             f" as it is already uploaded on {package_path}")
-                return
+        if not force_upload and _check_pex_version(resolved_fs, zip_file, package_path):
+            _logger.info(f"skip upload of current {zip_file}"
+                         f" as it is already uploaded on {package_path}")
+            return
+        elif resolved_fs.exists(package_path):
+            _logger.info(f"will replace current {zip_file} at {package_path}")
+            resolved_fs.rm(package_path)
 
     _logger.info(f"upload current {zip_file} to {package_path}")
 
     dir = os.path.dirname(package_path)
     if not resolved_fs.exists(dir):
         resolved_fs.mkdir(dir)
-    resolved_fs.put(zip_file, package_path)
+
+    if not _try_put_atomic(resolved_fs,

```

Comment:
This function returning a boolean to then return an error seems not very clean IMO and it's not very easy to understand the logic behind it.

Couldn't we move the L210-212 to L228 ? If we receive a FileExistsError + verify_expected_output is False then we raise, otherwise it is ok, the pex matches

Reply:
Ok indeed, I will simplify this